File: benchmarks/medium/M3-hashmap-stress/go/main.go
package main

import (
	"fmt"
	"math"
	"time"
)

const N = 5_000_000
const LOOKUPS = 1_000_000
const RUNS = 10

func main() {
	insertTimings := make([]float64, RUNS)
	lookupTimings := make([]float64, RUNS)

	fmt.Println("=== BENCHMARK: M3 - Hash Map Stress Test ===")
	fmt.Println("Language: Go")
	fmt.Println("Runtime:  Go")
	fmt.Println()

	var verifySum int64

	for run := 0; run < RUNS; run++ {
		// Insert phase
		start := time.Now()
		m := make(map[string]int64, N)
		for i := 0; i < N; i++ {
			key := fmt.Sprintf("user_%d", i)
			m[key] = int64(i)*31 + 7
		}
		insertElapsed := float64(time.Since(start).Nanoseconds()) / 1e6
		insertTimings[run] = insertElapsed

		// Lookup phase
		start = time.Now()
		state := uint32(42)
		var sum int64
		for j := 0; j < LOOKUPS; j++ {
			state = state*1664525 + 1013904223
			idx := uint64(state) % uint64(N)
			key := fmt.Sprintf("user_%d", idx)
			sum += m[key]
		}
		lookupElapsed := float64(time.Since(start).Nanoseconds()) / 1e6
		lookupTimings[run] = lookupElapsed

		verifySum = sum
		fmt.Printf("Run %2d: Insert: %8.2f ms | Lookup: %8.2f ms | Total: %8.2f ms\n",
			run+1, insertElapsed, lookupElapsed, insertElapsed+lookupElapsed)
	}

	fmt.Println()
	printStats("Insert", insertTimings)
	printStats("Lookup", lookupTimings)

	totalTimings := make([]float64, RUNS)
	for i := range RUNS {
		totalTimings[i] = insertTimings[i] + lookupTimings[i]
	}
	printStats("Total", totalTimings)

	fmt.Printf("Verification: sum = %d\n", verifySum)
}

func printStats(label string, timings []float64) {
	minT := timings[0]
	maxT := timings[0]
	sum := 0.0
	for _, t := range timings {
		sum += t
		if t < minT {
			minT = t
		}
		if t > maxT {
			maxT = t
		}
	}
	avg := sum / float64(len(timings))
	variance := 0.0
	for _, t := range timings {
		variance += (t - avg) * (t - avg)
	}
	variance /= float64(len(timings))
	stddev := math.Sqrt(variance)

	fmt.Printf("%s:\n", label)
	fmt.Printf("  Min:    %.2f ms\n", minT)
	fmt.Printf("  Avg:    %.2f ms\n", avg)
	fmt.Printf("  Max:    %.2f ms\n", maxT)
	fmt.Printf("  StdDev: %.2f ms\n", stddev)
	fmt.Println()
}

File: benchmarks/medium/M3-hashmap-stress/python/main.py
import math
import sys
import time


def main():
    N = 5_000_000
    LOOKUPS = 1_000_000
    RUNS = 10

    insert_timings = []
    lookup_timings = []

    print("=== BENCHMARK: M3 - Hash Map Stress Test ===")
    print("Language: Python")
    print(f"Runtime:  Python {sys.version.split()[0]}")
    print()

    verify_sum = 0

    for run in range(1, RUNS + 1):
        # Insert phase
        start = time.perf_counter()
        m = {}
        for i in range(N):
            m[f"user_{i}"] = i * 31 + 7
        insert_elapsed = (time.perf_counter() - start) * 1000
        insert_timings.append(insert_elapsed)

        # Lookup phase
        start = time.perf_counter()
        state = 42
        total = 0
        for _ in range(LOOKUPS):
            state = (state * 1664525 + 1013904223) & 0xFFFFFFFF
            idx = state % N
            total += m[f"user_{idx}"]
        lookup_elapsed = (time.perf_counter() - start) * 1000
        lookup_timings.append(lookup_elapsed)

        verify_sum = total
        print(f"Run {run:2d}: Insert: {insert_elapsed:8.2f} ms | Lookup: {lookup_elapsed:8.2f} ms | Total: {insert_elapsed + lookup_elapsed:8.2f} ms")

    print()
    print_stats("Insert", insert_timings)
    print_stats("Lookup", lookup_timings)

    total_timings = [i + l for i, l in zip(insert_timings, lookup_timings)]
    print_stats("Total", total_timings)

    print(f"Verification: sum = {verify_sum}")


def print_stats(label, timings):
    min_t = min(timings)
    avg_t = sum(timings) / len(timings)
    max_t = max(timings)
    variance = sum((t - avg_t) ** 2 for t in timings) / len(timings)
    stddev = math.sqrt(variance)

    print(f"{label}:")
    print(f"  Min:    {min_t:.2f} ms")
    print(f"  Avg:    {avg_t:.2f} ms")
    print(f"  Max:    {max_t:.2f} ms")
    print(f"  StdDev: {stddev:.2f} ms")
    print()


if __name__ == "__main__":
    main()
